### PR TITLE
Implement multiple target support for StreamJsonRpc

### PR DESCRIPTION
**Pull request by tinaschrepfer**
Implement multiple target support for StreamJsonRpc

### CODE REVIEW

File: src/StreamJsonRpc/JsonRpc.cs [hunk 1]
```diff
@@ -64,17 +63,15 @@ public class JsonRpc : IDisposableObservable
         /// </summary>
         private readonly Func<Task, object, JsonRpcMessage> handleInvocationTaskResultDelegate;
 
-        /// <summary>
-        /// A dictionary to map clr method names to <see cref="JsonRpcMethodAttribute" /> values.
-        /// </summary>
-        private readonly ReadOnlyDictionary<string, string> requestMethodToClrMethodMap;
+        private readonly List<Tuple<object, ReadOnlyDictionary<string, string>>> targetRequestMethodToClrMethodMap;
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: Can you add/update the xml doc comment for this member to clearly describe the contents of this collection?

File: src/StreamJsonRpc/JsonRpc.cs [hunk 2]
```diff
@@ -64,17 +63,15 @@ public class JsonRpc : IDisposableObservable
...
 
         private readonly CancellationTokenSource disposeCts = new CancellationTokenSource();
 
         private Task readLinesTask;
         private int nextId = 1;
         private bool disposed;
         private bool hasDisconnectedEventBeenRaised;
+        private bool startedListening = false;
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: FxCop whines when you initialize fields to their default values. Can you drop the `=false` from this field?

File: src/StreamJsonRpc/JsonRpc.cs [hunk 3]
```diff
@@ -230,10 +225,37 @@ public static JsonRpc Attach(Stream sendingStream, Stream receivingStream, objec
         }
 
         /// <summary>
+        /// Adds the specified targets as possible objects to invoke when incoming messages are received.
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: Either here or in the `<remarks>` section, can you explain the relationship these targets have with each other, and how order is significant?

File: src/StreamJsonRpc/JsonRpc.cs [hunk 4]
```diff
@@ -119,9 +115,8 @@ public JsonRpc(DelimitedMessageHandler messageHandler, object target = null)
             };
             this.JsonSerializer = new JsonSerializer();
 
-            this.requestMethodToClrMethodMap = this.callbackTarget != null
-                ? new ReadOnlyDictionary<string, string>(GetRequestMethodToClrMethodMap(this.callbackTarget))
-                : EmptyDictionary;
+            this.targetRequestMethodToClrMethodMap = new List<Tuple<object, ReadOnlyDictionary<string, string>>>();
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: nit: can you move this to a field initializer?

File: src/StreamJsonRpc/JsonRpc.cs [hunk 5]
```diff
@@ -230,10 +225,37 @@ public static JsonRpc Attach(Stream sendingStream, Stream receivingStream, objec
         }
 
         /// <summary>
+        /// Adds the specified targets as possible objects to invoke when incoming messages are received.
+        /// </summary>
+        /// <param name="targets">Targets to invoke when incoming messages are received.</param>
+        /// <remarks>This method must be called before JsonRpc starts listening for messages.</remarks>
+        public void AddLocalRpcTarget(params object[] targets)
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: Your method takes a plural argument yet your method name suggests singular.
We can rename the method, or take a singular parameter. I'm leaning toward a singular parameter, myself, based on my belief that most callers will have just one target to "add" so we'll avoid a loop and an array allocation by just taking one. And callers would then be responsible to avoid calling it with a null argument.

File: src/StreamJsonRpc/JsonRpc.cs [hunk 6]
```diff
@@ -230,10 +225,37 @@ public static JsonRpc Attach(Stream sendingStream, Stream receivingStream, objec
...
+        {
+            Requires.NotNull(targets, nameof(targets));
+
+            if (this.startedListening)
+            {
+                throw new InvalidOperationException("Cannot attach additional targets once JsonRpc has started listening for messages.");
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: Please move the message to `Strings.resx` so it can be localized.
Also, you can reduce these 4 lines to just one by calling `Verify.Operation`

File: src/StreamJsonRpc/JsonRpc.cs [hunk 7]
```diff
@@ -230,10 +225,37 @@ public static JsonRpc Attach(Stream sendingStream, Stream receivingStream, objec
...
+            }
+
+            foreach (var target in targets)
+            {
+                if (target != null)
+                {
+                    this.targetRequestMethodToClrMethodMap.Add(new Tuple<object, ReadOnlyDictionary<string, string>>(
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: Tip: you can use `Tuple.Create(` instead of `new` here to avoid having to specify all the generic type arguments since the inferred type arguments will be equivalent.

File: src/StreamJsonRpc.Tests/JsonRpcTests.cs [hunk 8]
```diff
@@ -558,6 +558,44 @@ await invokeTask.NoThrowAwaitable(); // may or may not throw due to cancellation
         await Assert.ThrowsAsync<RemoteMethodNotFoundException>(() => this.clientRpc.InvokeAsync<string>(nameof(Server.TestInvalidMethod)));
     }
 
+    [Fact]
+    public void AddLocalRpcTarget_ExceptionThrownWhenRpcHasStartedListening()
+    {
+        Assert.Throws<InvalidOperationException>(() => this.clientRpc.AddLocalRpcTarget(new AdditionalServerTargetOne()));
```
Comment (AArnott) [inline src/StreamJsonRpc.Tests/JsonRpcTests.cs]: Can we also test a null `object[]` and an array with null elements?

File: src/StreamJsonRpc/JsonRpc.cs (hunk 2 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc/JsonRpc.cs]: Done.

File: src/StreamJsonRpc/JsonRpc.cs (hunk 4 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc/JsonRpc.cs]: Done.

File: src/StreamJsonRpc/JsonRpc.cs (hunk 3 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc/JsonRpc.cs]: Done.

File: src/StreamJsonRpc/JsonRpc.cs (hunk 5 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc/JsonRpc.cs]: Done

File: src/StreamJsonRpc/JsonRpc.cs (hunk 6 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc/JsonRpc.cs]: Done.

File: src/StreamJsonRpc/JsonRpc.cs (hunk 7 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc/JsonRpc.cs]: Done.

File: src/StreamJsonRpc.Tests/JsonRpcTests.cs (hunk 8 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc.Tests/JsonRpcTests.cs]: Done.

File: src/StreamJsonRpc/JsonRpc.cs [hunk 9]
```diff
@@ -230,10 +230,26 @@ public static JsonRpc Attach(Stream sendingStream, Stream receivingStream, objec
...
         /// <summary>
+        /// Adds the specified target as possible object to invoke when incoming messages are received.  The target object
+        /// should not inherit from each other and are invoked in the order which they are added.
+        /// </summary>
+        /// <param name="target">Target to invoke when incoming messages are received.</param>
+        /// <remarks>This method must be called before JsonRpc starts listening for messages.</remarks>
+        public void AddLocalRpcTarget(object target)
+        {
+            Requires.NotNull(target, nameof(target));
+            Verify.Operation(!this.startedListening, Resources.BothReadableWritableAreNull);
```
Comment (IlyaBiryukov) [inline src/StreamJsonRpc/JsonRpc.cs]: Copy-pasted? Should be a different error message I think.

**Comment by AArnott** [comment]
@tinaschrepfer just to call out that Ilya's last couple of comments are still pending resolution.
Also: what version of VS do you anticipate consuming these changes in?

File: src/StreamJsonRpc/JsonRpc.cs (hunk 9 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc/JsonRpc.cs]: That's so weird.  I swear when I wrote this it was the right message.  I'll change it.

**Comment by tinaschrepfer** [comment]
@AArnott, comments addressed.  Our feature is targeting Dev16.  However, without these changes in place for StreamJsonRpc, I can't consume the new APIs for custom messages.  If you could push these changes into vscore or something, I can always build against that branch.  Thanks!

File: src/StreamJsonRpc/JsonRpc.cs [hunk 10]
```diff
@@ -773,13 +773,13 @@ private static JsonRpcMessage CreateError(JToken id, Exception exception)
                 foreach (var targetMap in this.targetRequestMethodToClrMethodMap)
                 {
                     targetMethod = new TargetMethod(request, targetMap.Item1, jsonSerializer, targetMap.Item2);
-                    if (targetMethod.IsFound)
+                    if (targetMethod != null && targetMethod.IsFound)
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: We don't need a `targetMethod != null` check here because you just assigned it with `new`, which cannot return null.

File: src/StreamJsonRpc/JsonRpc.cs [hunk 11]
```diff
@@ -773,13 +773,13 @@ private static JsonRpcMessage CreateError(JToken id, Exception exception)
...
                     {
                         break;
                     }
                 }
 
-                if (!targetMethod.IsFound)
+                if (targetMethod == null || !targetMethod.IsFound)
```
Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: Thanks for this change. Can we add a test to prove we never throw NRE when `targetMethod==null`, since evidently we are missing one?

File: src/StreamJsonRpc/JsonRpc.cs (hunk 10 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc/JsonRpc.cs]: Whoops, wasn't thinking.  Removed.

File: src/StreamJsonRpc/JsonRpc.cs (hunk 11 above)
Comment (tinaschrepfer) [inline src/StreamJsonRpc/JsonRpc.cs]: I actually don't think targetMethod can ever be null.  The only scenario where that's possible is if targetRequestMethodToClrMethodMap is empty.  But we already check for that beforehand and return a different error.  In any case, I added a test to test that a remote target exception is thrown if we call InvokeAsync with no target object.

Comment (AArnott) [inline src/StreamJsonRpc/JsonRpc.cs]: doh! I guess we already had such a test. I'll remove the extra and merge your PR. Thanks!
